File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.java
import java.util.*;

class Solution {
    public int solution(String[] board) {
        int[] r = new int[2];
        
        for (int j = 0; j < board.length; j++) {
            for (int i = 0; i < board[0].length(); i++) {
                if (board[j].charAt(i) == 'R') {
                    r = new int[]{j, i};
                }
            }
        }
        
        Deque<Robot> q = new LinkedList<>();
        q.add(new Robot(r[0], r[1], 0));
        
        int[] dy = {0,0,1,-1};
        int[] dx = {1,-1,0,0};
        boolean[][] visited = new boolean[board.length][board[0].length()];
        
        while (!q.isEmpty()) {
            Robot robot = q.pollFirst();
            int count = robot.count + 1;
            
            for (int i = 0; i < 4; i++) {
                int ny = robot.y;
                int nx = robot.x;
                while(0 <= ny + dy[i] && ny + dy[i] < board.length
                      && 0 <= nx + dx[i] && nx + dx[i] < board[0].length()
                      && board[ny+dy[i]].charAt(nx + dx[i]) != 'D') {
                        ny += dy[i];
                        nx += dx[i];
                    }
                 
                if (board[ny].charAt(nx) == 'G') {
                    return count;
                }
                
                if (!visited[ny][nx]) {
                    visited[ny][nx] = true;
                    q.add(new Robot(ny, nx, count));
                }
            }
        }
        
        return -1;
    }
    
    static class Robot {
        int y;
        int x; 
        int count;
        
        public Robot(int y, int x, int count) {
            this.y = y;
            this.x = x;
            this.count = count;
        }
    }
}

File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.py
from collections import deque

def solution(board):
    visited = [[False] * len(board[0]) for _ in range(len(board))]
    
    dy = [0,0,1,-1]
    dx = [1,-1,0,0]
    
    q = deque()
    
    for j in range(len(board)):
        for i in range(len(board[0])):
            if board[j][i] == "R":
                visited[j][i] = True
                q.append((j, i, 0))
                
                while q:
                    y, x, count = q.popleft()
                    
                    for k in range(4):
                        ny = y
                        nx = x
                        
                        while 0 <= ny + dy[k] < len(board) and 0 <= nx + dx[k] < len(board[0]) and board[ny+dy[k]][nx+dx[k]] != "D":
                            ny += dy[k]
                            nx += dx[k]
                        
                        if not visited[ny][nx]:
                            if board[ny][nx] == "G":
                                return count + 1
                            visited[ny][nx] = True
                            q.append((ny, nx, count+1))
    return -1